compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -356,11 +361,11 @@ fn check_for_bindings_named_same_as_variants(
 }
 
 /// Checks for common cases of "catchall" patterns that may not be intended as such.
-fn pat_is_catchall(pat: &DeconstructedPat<'_, '_>) -> bool {
+fn pat_is_catchall(pat: &DeconstructedPat<'_, MatchCheckCtxt<'_>>) -> bool {
     use Constructor::*;
     match pat.ctor() {
         Wildcard => true,
-        Single => pat.iter_fields().all(|pat| pat_is_catchall(pat)),
+        Single | Tuple(_) | Ref(_) | BoxPat(_) => pat.iter_fields().all(|pat| pat_is_catchall(pat)),
```

Comment:
Is this a bugfix?

This seems like it might be a source of performance impact.

Reply:
No, this is because I split the `Single` case into more specific subcases so that I could abstract it all in a trait. I suspected this could be the source of perf degradation but my local measurements don't show much impact from this.
The problem of course is that my local measurements don't match what `rust-timer` reports here so you could be right. When I get back to this I'll measure intermediate commits with `rust-timer`.